teamhanko/hanko
# fix: persisted passcode timestamps

# Description

Time.Now() uses local time but timezone information is lost on persisting because pop timestamp columns are without timezone. On retrieval from the DB the original timestamp is not wholly recoverable and leads to erroneous comparisons (e.g. passcode expiry check). 

# Implementation

The changes solve the problem by explicitly using UTC both on save and comparison.

# Test

How to test:

1. On the `main` branch, run the backend using a low passcode TTL, e.g. in the config set:
```
passcode:
  ttl: 1
```
2. Run a [mailslurper](https://hub.docker.com/r/marcopas/docker-mailslurper) 
3. Initialize a passcode, check mailslurper, finalize passcode with code
4. Note how the finalization succeeds even though the 1 second TTL has long since elapsed

Repeat 1-3 on this branch, finalization should then return an error (passcode expired).
